### ISSUE TITLE
Feedback from UXnorge article

### ISSUE THREAD

**Issue by Febakke** [digdir/designsystemet#3379]
Yesterday UXnorge posted an article on improvements we should do to Designsystemet concerning Accessibility. I created this issue so that we can investigate what points should have issues, if there is anything need to research more and what should be discussed more before fixing.

We want to thank Kristian Munter Simonsen for contributing to this design system by testing it for accessibility issues. You can read the full article on [UXNorge: Universell utforming i Designsystemet](https://uxnorge.no/universell-utforming-i-designsystemet-en-investering-for-fremtiden/) 👏

## Documentation
We have not prioritized documentation leading up to the release. This will be a priority moving forward. 

### Tabs
Here are the comments, this should be taken into account when writing guidelines for tabs

> Det er ingen spesifiseringer rundt tilgjengelighet ved bruk av tab og tablist elementene. Dette er elementer vi i Inklud ofte ser at brukere av hjelpemiddelverktøy har utfordringer med. Samtidig opplever vi også ofte at slike elementer blir feilaktig rapportert som feil da man ikke kan tabbe til alternativene i tablisten. Det ville vært nyttig å forklare hvilke forventninger man skal ha for slike komplekse elementer. Bruk av tabindex burde også vurderes som et alternativ for å påse at alle brukere får tilgang til tab-elementene. 

### Input types
The comment here was that we implied that it was up to the developer what input types should be used when setting up a input field. But according to [1.3.5 Identify input purpose](https://www.w3.org/WAI/WCAG21/Understanding/identify-input-purpose.html) this is mandatory to use correct input type. ~~We will change our documentation now to be less vague about this.~~ I am not sure that this is correct. We will document what our conclusion is in #3373

## Patterns
The collaboration on Patterns are time consuming and requires collaboration and resources from multiple government agencies. It might therefor be some waiting time for all the patterns we want to look at are finished. We hope that we can have more resources working on this in the future. We encourage everyone to vote, comment and create discussions on our discussions page regarding patterns. This will influence how we prioritize.

[Pattern discussions](https://github.com/digdir/designsystemet/discussions/categories/samarbeid-m%C3%B8nster-for-innbyggertjenester)

Updated patterns with feedback from article
[Login](https://github.com/digdir/designsystemet/discussions/1676)
[Confirmation page on forms](https://github.com/digdir/designsystemet/discussions/3381)

### Link
Article proposes that we make `href` required for our link component. This makes sense as a link is only a link if this is present. But `href` is not required when using our Link component because some libraries or frameworks have their own links, in these cases you would use `asChild` on our link to get correct styling. Because of this we cant make `href` required as it would break this functionality. 

### Role over aria
The article says that aria-attributes have better assistive technology support then aria-roles (Roles can be seen as a bundle of multiple aria-attributes) 

[UUTilsynets article on 4.1.3 Statusbeskjeder](https://www.uutilsynet.no/wcag-standarden/413-statusbeskjeder-niva-aa/152) uses both roles and attributes as example of handling this issue. 

[WCAG: 4.1.3 Status Message](https://www.w3.org/WAI/WCAG22/Understanding/status-messages.html) techniques also uses roles as examples on how to solve. 

There is no warning on the [documentation page for the role alert](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/alert_role)

[This test shows some less support for JAWS but is 4-5 years old](https://a11ysupport.io/tech/aria/alert_role)
[This article says the opposite](https://www.accessibility-developer-guide.com/examples/sensible-aria-usage/alert/#live-regions-an-alternative)

How is a developer to know what is the right choice? We went with roles when working on the systemalerts patterns and tried to be consistent in that choice. Our understanding was that both were supported. We need to know what the issue is and who it affects, is it just `role="alert"` or is this also an issue with others aria-roles? 

I have started documenting this here: #3383

### Error message read twice
When testing I found that error message when using max character function on text fields are read twice. (VoiceOver and Chrome) 
But I think the feedback is that we have error message connected to our input and the error message is accessible beneath the input. So if a screen reader goes element to element it will get the error message read both when the field has focus and when the error message beneath the field is in focus. 

Issue here: #3374

## Examples with WCAG errors
We have testing pages on our Storybook that is used for technical testing. For instance see that all components have correct height. In the future we will build our own Storefront for both components and examples. Storybook will be hidden and used as a developer tool. 

We are discussing if we should hide the test pages while we wait for our own storefront or fix it in some other way.

**Comment by Mokkatroffen** [comment]
Hello! Very happy to see that the issues highlighted has been documented here on GitHub 👍 
@Febakke contacted me on the Designsystemet Slack channel and asked about the role/aria definition issues note. 

**TLDR:** The role definition of status messages is not the worst offence when using roles in my book. Due to factors regarding implementation and adopters/implementers understanding of the Designsystemet components and complex accessibility aspects like dynamic announcement, it is my opinion that one should use aria-live in combination with other aria attributes like atomic over role alert and status. It also makes it less likely that status messages will be announced each time a page is loaded for screen reader users. The use of role is generally more reprehensible in other instances. 

**Preface:**
Let me elaborate my knowledge of the issue and you can yourself evaluate how you want to handle it (or hire me for a few hours to partake in some additional discussions.
But let me first preface some stuff that is quite important that we are all on the same page about. 
What constitutes a breach of a success criterion depends and where and when you are (in time and space)… What I mean is that in Norway it is the national audit agency which dictates what constitutes a breach of requirement and what does not. Uutilsynet uses WCAG and its techniques to formulate their own interpretations of what they test for, which can change and be adjusted over time. 
What is legal and what is a robust and good user experience is overlapping, but you can be completely legal in many aspects of accessibility of ICT and not have an accessible ICT solution. Beyond the fact that there is no current legal requirement for font-size or size of interactive elements we must afford the effort of aiming to create something that works well and is robust, especially across assistive technologies, browsers and operating systems. 

**Role referred to as within the requirements**
Yes, uutilsynet and WCAG specifically reflects roles as a sufficient way to comply with the requirement. But note that both of them also include properties and aria as well. 

**Role and ARIA**
Once one starts using roles one should know all the values that are being set in the accessibility tree. One of the dangers with using role alert and status is that one does not acquire or can expect users to properly understand how one gets the role to behave exactly how one intends/wants to. 

**Support issues**
Yes, it is not the most updated test results. I would gladly preform some additional testing if requested 😉 

[Aria-live support](https://a11ysupport.io/tech/aria/aria-live_attribute):

![Image](https://github.com/user-attachments/assets/e2dc41cf-ef0e-4fc6-88a9-dc65110bc5fb)

 
[Role alert support](https://a11ysupport.io/tech/aria/alert_role):

![Image](https://github.com/user-attachments/assets/ad4bb5a0-cf3b-4562-906c-2f7c1c6d5ee2)

 
I don’t believe the big issue here is the support alone. I have not recently been able to check the current state of role alert vs aria-live assertive. 
Support issues also hold true across native HTML5 elements, like <dl> <dt>, <dd> etc. In the <dl> etc. case it is better to use more adopted elements like <li> or <ol>. 
When one gets into more complicated webcontent, like chatbots it is important to be able to segment each attribute, like with chatbots. If only roles are available that work will become very difficult. 

**Polite option**
To have the knowledge that role alert = assertive announcement and that there is an alternative role status = polite announcement is to expect a lot. It might just be a personal opinion, but I believe that one should explicitly choose to define live content like status messages.
If not you run the risk of a status message or error message being used across pages and announced every time the pages loads. It might become more problematic when users might introduce dynamic content within the role definition. This can happened when one uses aria-live as well, don’t get me wrong, but once you allow and require users to take a stance on whether this content should be read out loud (aria-live or not), and what should happened if there is dynamic content within the live region you enable users of Desingsystemet a finer degree of nuance and understanding. 

**Personally damaged by rule #1 of use of aria**
I do quite a lot of audits and role is always a red flag when it comes to web accessibility. After further inspection it seems like one has in general avoided use of aria when possible. One instance of where it should be defined is for the search component, one of the few roles which cannot be defined through standard HTML5 elements. Please add a <form> element for the role search and introduce role=search to the <form> element. Optionally inform about the fact that sitewide search functionality should be defined with a <form> with role search.  

**Bonus – Contrast issue**
Placeholder text also needs to adhere to contrast requirements. The placeholder text in the screenshot bellow is not of sufficient contrast (less than 4.5:1). This is just another good reason to not use placeholder:

![Image](https://github.com/user-attachments/assets/c12723ad-719c-4279-8ec6-65a45bb579ce)

**Comment by mrosvik** [comment]
We have created issues for the items that need to be addressed and will therefore close this issue. @Febakke will do a final review before it is closed.

**Comment by mimarz** [comment]
Everything here has now been addressed in other issues, so we are closing this one.
